src/Razor/src/Microsoft.AspNetCore.Razor.ProjectEngineHost/Utilities/Checksum.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.AspNetCore.Razor.Utilities;
+
+internal sealed partial class Checksum : IEquatable<Checksum?>
+{
+    private const int HashSize = 20;
+
+    public static readonly Checksum Null = new(default);
+
+    public readonly HashData Data;
+
+    public Checksum(HashData data)
+        => Data = data;
+
+    public static Checksum From(ReadOnlySpan<byte> source)
+    {
+        if (source.Length == 0)
+        {
+            return Null;
+        }
+
+        if (source.Length < HashSize)
+        {
+            throw new ArgumentException($"{nameof(source)} must be equal to or larger than the hash size: {HashSize}", nameof(source));
```

Comment:
Why is it ok if the input is _bigger_ than the expected hash size?

Reply:
`MemoryMarshal.TryRead(...)` needs to be passed a span at least as large as `HashData` or it'll return false. This just catches that condition early and throws a better exception.

Reply:
I guess my concern is better stated as: if we don't have a hash exactly the length we were expecting, is it even something we can interpret as a hash?